keps/sig-apps/3715-elastic-indexed-job/README.md

```diff
@@ -0,0 +1,934 @@
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Allow mutating `spec.completions` for `Indexed` job iff the updated value is equal to `spec.parallelism`.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Change the existing behavior of `NonIndexed` mode.
+- Change the existing behavior of `Indexed` mode for jobs that never mutate.
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+I have a batch workload that requires two sets of pods: a manager and a group of workers that can scale up or down. 
+The manager communicates with the workers to manage their load; examples of such workloads include distributed HPC 
+and districuted ML training. Examples of frameworks include Horovod with MPI, Spark and Ray.
+
+This workload can be modeled as two `Indexed` Jobs, one for the manager and one for the workers. A 
+headless service is created to set up stable hostnames for the worker pods . The workers job is
+scaled up/down by updating `spec.completions` and `spec.parallelism` in tandem.
+
+The success semantics of these workloads are tied to the manager, not the workers. However, in
+such workloads, when the job is finished the workers are either deleted directly (effectively scaled down to 0)
+or they potentially exit on their own after communicating their final status to the manager who decides the final status
+of the whole workload.
+
+
+<!--
+
+### Notes/Constraints/Caveats (Optional)
+
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+There is a concern that changing completion semantics on the fly can be subject to
+race conditions with declaring the job finished. However, after a closer look, this 
+seems to not be an issue because the job controller uses Update when setting the
+finished status on the job object, which is subject to object modified errors if the spec
+changes at the same time.
+
+<!--
```

Comment:
Let's also add that `completions` will be immutable once a job has a finished condition.